### PR TITLE
Optimisations for binary operations broadcast. Algorithm improvements.

### PR DESCRIPTION
**Pull request by vladislav-volkov**
This fix contains optimizations for tensor binary operations broadcasting.

### CODE REVIEW

File: ngraph/src/ngraph/coordinate_transform.cpp [hunk 1]
```diff
@@ -29,6 +29,57 @@
...
 
+namespace
+{
+    Strides default_strides(size_t n_axes) { return Strides(n_axes, 1); }
+    CoordinateDiff default_padding(size_t n_axes) { return CoordinateDiff(n_axes, 0); }
+    AxisVector default_axis_order(size_t n_axes)
+    {
+        AxisVector result(n_axes);
+        size_t n = 0;
+        std::generate(result.begin(), result.end(), [&n]() -> size_t { return n++; });
```
Comment (postrational) [inline ngraph/src/ngraph/coordinate_transform.cpp:40]: May be simpler (faster?) to use `std::iota` here.

File: ngraph/src/ngraph/coordinate_transform.hpp [hunk 2]
```diff
@@ -24,9 +24,48 @@
 
 namespace ngraph
 {
-    class NGRAPH_API CoordinateTransform
+    class NGRAPH_API CoordinateIterator
```
Comment (postrational) [inline ngraph/src/ngraph/coordinate_transform.hpp:27]: Please add documentation which explains when to use this class.

File: ngraph/src/ngraph/coordinate_transform.hpp [hunk 3]
```diff
@@ -24,9 +24,48 @@
...
+        static const CoordinateIterator& end();
+
+    private:
+        Shape m_target_shape;
+        Coordinate m_coordinate;
+        bool m_oob;
+        bool m_empty;
+    };
+
+    class NGRAPH_API CoordinateTransformBasic
```
Comment (postrational) [inline ngraph/src/ngraph/coordinate_transform.hpp:48]: Please add documentation which explains when to use this class.

File: ngraph/src/ngraph/coordinate_transform.hpp [hunk 4]
```diff
@@ -24,9 +24,48 @@
...
+        size_t index(const Coordinate& c) const noexcept;
+
+        CoordinateIterator begin() const noexcept;
+        const CoordinateIterator& end() const noexcept;
+
+    protected:
+        Shape m_source_shape;
+    };
+
+    class NGRAPH_API CoordinateTransform : protected CoordinateTransformBasic
```
Comment (postrational) [inline ngraph/src/ngraph/coordinate_transform.hpp:64]: Please add documentation which explains when to use this class.

File: ngraph/src/ngraph/coordinate_transform.hpp [hunk 5]
```diff
@@ -24,9 +24,48 @@
...
     public:
+        CoordinateIterator(const Shape& target_shape, bool is_end = false);
+
+        void operator++();
+        CoordinateIterator operator++(int);
+        void operator+=(size_t n);
+        const Coordinate& operator*() const noexcept;
+        bool operator!=(const CoordinateIterator& it) const noexcept;
+        bool operator==(const CoordinateIterator& it) const noexcept;
+        void advance(size_t axis) noexcept;
```
Comment (postrational) [inline ngraph/src/ngraph/coordinate_transform.hpp:38]: Please add documentation string.

File: ngraph/src/ngraph/coordinate_transform.cpp (hunk 1 above)
Comment (vladislav-volkov) [reply ngraph/src/ngraph/coordinate_transform.cpp:40]: I kept original code unchanged. Ok, I'll fix code according your comment.

File: ngraph/src/ngraph/coordinate_transform.hpp (hunk 2 above)
Comment (vladislav-volkov) [reply ngraph/src/ngraph/coordinate_transform.hpp:32]: fixed

File: ngraph/src/ngraph/coordinate_transform.hpp (hunk 3 above)
Comment (vladislav-volkov) [reply ngraph/src/ngraph/coordinate_transform.hpp:79]: fixed

File: ngraph/src/ngraph/coordinate_transform.hpp (hunk 4 above)
Comment (vladislav-volkov) [reply ngraph/src/ngraph/coordinate_transform.hpp:103]: fixed

File: ngraph/src/ngraph/coordinate_transform.hpp (hunk 5 above)
Comment (vladislav-volkov) [reply ngraph/src/ngraph/coordinate_transform.hpp:63]: fixed

**Comment by ilya-lavrenov** [comment]
@vladislav-volkov do we have any performance numbers?

**Comment by vladislav-volkov** [comment]
> @vladislav-volkov do we have any performance numbers?

Yes I have the following results.

```1000000 additions of 2 tensors with different dimensions

Test 0
{4} x {}
        original    with optimisations  speed-up
real    0m1,203s    0m0,091s            13.2 times
user    0m1,199s    0m0,091s
sys     0m0,004s    0m0,000s

Test 1
{3, 2, 1} x {1, 6}
        original   with optimisations   speed-up
real    0m8,315s   0m0,183s             45.4 times
user    0m8,315s   0m0,183s
sys     0m0,000s   0m0,000s

Test 2
{2, 12, 1, 1, 1} x {1, 1, 1, 8}
        original    with optimisations  speed-up
real    0m44,200s   0m0,472s            93.6 times
user    0m44,201s   0m0,472s
sys     0m0,001s    0m0,004s

Test 3
{3, 24, 1, 1, 1, 1, 1} x {1, 1, 1, 1, 1, 64}
        original    with optimisations  speed-up
real    19m33,473s  0m1,779s            659.6 times
user    19m33,477s  0m1,779s
sys     0m0,000s    0m0,000s

Test 4
{2, 12, 1, 2, 8} x {1, 1, 2, 8}
        original    with optimisations  speed-up
real    2m1,054s    0m0,474s            255.4 times
user    2m1,054s    0m0,474s
sys     0m0,000s    0m0,000s


Test 5
{3, 2, 8} x {3, 2, 8}
        original     with optimisations  speed-up
real    0m15,584s    0m0,040s            389.6 times
user    0m15,584s    0m0,040s
sys     0m0,000s     0m0,000s

Test 6
{1, 3, 2, 8, 1, 1, 1} x {1, 3, 2, 8, 1, 1, 1}
        original     with optimisations  speed-up
real    0m19,195s    0m0,050s            383.9 times
user    0m19,191s    0m0,050s
sys     0m0,004s     0m0,000s
```

**Comment by ilyachur** [comment]
@postrational All your comments were fixed. I will merge this PR.

**Pull request merged**